core/trino-main/src/main/java/io/trino/sql/planner/JsonPathInvocationContext.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.collect.cache.NonEvictableCache;
+import io.trino.metadata.BoundSignature;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.OperatorNotFoundException;
+import io.trino.metadata.ResolvedFunction;
+import io.trino.spi.function.OperatorType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeManager;
+import io.trino.type.TypeCoercion;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCache;
+import static io.trino.sql.planner.JsonPathInvocationContext.Invocation.RESOLUTION_ERROR;
+import static io.trino.sql.planner.JsonPathInvocationContext.Invocation.invocation;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Per-thread cache for resolved operators and coercions used by the JsonPathEvaluator.
```

Comment:
Yes, it was supposed to be per driver.
We did not do global caching to avoid many threads synchronising on this cache. Also, so it was suggested by @martint.